### ISSUE TITLE
[Go_router] Add current and next uri to redirects

### ISSUE THREAD

**Issue by cedvdb** [flutter/flutter#141984]
### Use case

I want to access the current uri from inside a redirect function. While it's possible today, it is quite ugly as (you essentially need to ignore a linter by predeclaring a router instance to use it inside the redirect method and the way you access the current uri is equally ugly).

Let's take this common scenario, of unprotected and protected routes:

 - /products/:id (unprotected)
 - /products/:id/edit ( requires account)
 - /suppliers/:id (unprotected)
 - /suppliers/:id/edit (requires account)
 - /create-account (requires authentification)
 - /sign-in

Due to how go_router redirection work, when an unauthenticated user is on /products/:id and access /products/:id/edit he will be redirected to /create-account then finally to /sign-in. The page stack will only contain the /sign-in page. Since there is only one page on the stack, the Scaffold will not display a BackButton, rendering the user stuck on the /sign-in page.

As of today it is possible to display a BackButton by adding it manually to the scaffold and somehow get access to the current uri in the redirect method to redirect to :  `/sign-in?onSuccess=the-url-user-wanted-to-access&onNavBack=previous-url` (note onNavBack query param) to allow navigation to the previous page.

### Proposal

The proposal is to give the `current` state and the `next` state to the redirect function.  Currently only the target state is passed.

IE: transform

```dart
/// The signature of the redirect callback.
typedef GoRouterRedirect = FutureOr<String?> Function(
    BuildContext context, GoRouterState state);
```

into

```dart
typedef GoRouterRedirect = FutureOr<String?> Function(
    BuildContext context, 
    GoRouterState current,
    GoRouteState next,
);
```

**Comment by cedvdb** [comment]
@chunhtai 

> Do you have some example to show case how would one use this new API and the expectation?

Yes I can since I wrap go router currently to add this functionality. 

> It sounds like you want to redirect by push?

I've explored many solutions for my use case including redirecting by push (as I said I wrap go router). I'd advice in threading carrefully if you end up implementing that proposal and thing of different scenarios where there are multiple redirects after another. I tried with this method but couldn't manage to make it work gracefully enough for my taste but I cannot give details on top of my head

That being said here is how I achieved what I wanted:

  => on redirect: add success uri and nav back uri to query parameters. The success uri being the `next` uri in my proposal since it's the one that was originally accessed before redirect. The nav back uri is the `current` uri in my proposal.


Here is a copy paste:

```

  Uri addReturnUrlsQueryParam({
    // the uri we are currently
    required Uri? current,
    // the uri we are redirected to next
    required Uri redirect,
    // the uri we wanted to be at
    required Uri target,
    // the uri where we go to on nav back
    required Uri? onNavBack,
  }) {
    // if the link already contains redirectTo use those instead
    Map<String, String> queryParams = {
      ...(current?.queryParameters ?? {}),
      ...target.queryParameters
    };

    final currentOnSuccess = queryParams[_uriOnSuccessKey];
    final currentOnNavBack = queryParams[_uriOnNavBackKey];
    if (currentOnSuccess == null) {
      queryParams = {
        ...queryParams,
        _uriOnSuccessKey: Uri.encodeComponent(target.path),
      };
    }
    if (currentOnNavBack == null && onNavBack != null) {
      // we cant have success and nav back be the same
      if (Uri.encodeComponent(onNavBack.path) !=
          queryParams[_uriOnSuccessKey]) {
        queryParams = {
          ...queryParams,
          _uriOnNavBackKey: Uri.encodeComponent(onNavBack.path),
        };
      }
    }

    redirect = redirect.replace(queryParameters: queryParams);
    return redirect;
  }
```

**Comment by chunhtai** [comment]
Design doc to add this feature http://flutter.dev/go/go-router-redirect

**Comment by nerder** [comment]
To add context to this, there are other scenarios where this might be useful.

Here is my current use-case.

In our app we are currently leveraging Stripe Checkout (no-code solution) for letting users setup a payment method. In order to do that we open a browser for the user as soon as the backend has been able to generate the checkout session URL. That session also contains a `cancelUrl` and a `successUrl` callbacks.

Those callback are configured as deep-links so that the app is opened when the user either complete the setup of the payment method or cancel.

Our intended behaviour in this case is to prevent any navigation and simply display a `SnackBar` in case of success. This is currently not possible to achieve with the current `redirect` method because by not having access to the `current` path we can't "cancel" the redirect by returning the same path, nor we can return `null` because this will end up in the `errorPageBuilder` as the route from the deep-link is not registered.

In our scenario this is an issue becase the navigation from the deep link can be triggered from multiple pages in the app, so we can't simply hard-code the redirection but we need to know where the user were to redirect them correctly.

Hopefully this adds some more priority to the issue.
